File: src/algorithms/go/quickSort.go
package main

func quickSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	pivot := arr[len(arr)/2]
	var left, middle, right []int

	for _, x := range arr {
		if x < pivot {
			left = append(left, x)
		} else if x == pivot {
			middle = append(middle, x)
		} else {
			right = append(right, x)
		}
	}

	left = quickSort(left)
	right = quickSort(right)

	return append(append(left, middle...), right...)
}

File: src/algorithms/cpp/quickSort.cpp
#include <iostream>
#include <vector>

std::vector<int> quickSort(std::vector<int> &arr)
{
    if (arr.size() <= 1)
    {
        return arr;
    }

    int pivot = arr[arr.size() / 2];
    std::vector<int> left, middle, right;

    for (int x : arr)
    {
        if (x < pivot)
        {
            left.push_back(x);
        }
        else if (x == pivot)
        {
            middle.push_back(x);
        }
        else
        {
            right.push_back(x);
        }
    }

    left = quickSort(left);
    right = quickSort(right);

    std::vector<int> sortedArr;
    sortedArr.reserve(left.size() + middle.size() + right.size());
    sortedArr.insert(sortedArr.end(), left.begin(), left.end());
    sortedArr.insert(sortedArr.end(), middle.begin(), middle.end());
    sortedArr.insert(sortedArr.end(), right.begin(), right.end());

    return sortedArr;
}